### PR TITLE
Be graceful about unsupported gradient types.

### PR DESCRIPTION
**Pull request by whilo**
It is helpful to know when the gradient fails. I really like your work on wrapping tensorflow, cortex is more for users, it does not provide the same auto-differentiation toolbox. I am exploring variational objects like VAEs and my impression from browsing the cortex codebase a bit is that it will be difficult to decompose and compose it for research. It seems like a layer is a central abstraction. Why exactly is the gradient not working automatically for tensorflow on the JVM? 

I am a bit torn as I am more interested in ML for my thesis than in reimplementing an auto-differentiation engine, so I might go back to Python for my master thesis if this does not work out. On the other hand I am exploring composition with Bayesian Program Learning for it and there Anglican is pretty nice. Do you think you can support me to get my VAE stuff going quickly? Do you have a communication channel? I will try to implement the missing gradients then.

In case of a complete graph for auto-differentiation in Clojure I would also like if tensorflow would just be a backend in the long-run, so we could leverage the uncomplicate libs from Dragan Djuric to have a full high-perf compilation pipeline for researchers in Clojure. That is a long-term idea though.

### CODE REVIEW

**Comment by kieranbrowne** [comment]
Hi @whilo,
Thanks for the PR. 
VAEs should work theoretically, but we'll probably need to fill in the missing gradients for any ops your want to use. Gradient computation is not currently working for Java's TF api because it hasn't been implemented yet.
I wrote the gradient descent algorithm in this lib but have only set it up for operations that I've been using so far. It also might be a naive implementation.
Can you give me some more details about the VAE you're trying to build, I'll try to help you get it set up.
I don't currently have a communication channel setup but I'm open to the idea.
Thanks again,
Kieran

**Comment by whilo** [comment]
I am a bit frustrated with the take on scientific computing in Clojure. I don't have the feeling I have composable building blocks which expose the proper low-level knobs for optimization. `core.matrix` is nice, but it does not have a fast backend (theano was 5x times faster on CPU with the clatrix backend 2 years ago on an RBM on MNIST). Incanter was a nice attempt, but it feels like a half-finished framework with some simple methods that are not as composable as Clojure is in general (e.g. their dataframe abstraction) and is inactive. 

In general the JVM is weak in scientific computation, although it shouldn't be. I think Dragan Djuric is doing the right thing about considering performance first and exposing that through a nice mapping to LAPACK. This performance advantage will be an important benefit that is not recoverable if simplifying abstractions a la numpy (core.matrix) are introduced. It is also important to stay competitive against Julia. Beating Python/R is easy.

 I also think Clojurians should help the JVM here (not just do something for Clojure). Scientific computing is very important to many JVM users. In that regard I think supporting a legacy Python stack is a first step to avoid uphill battles: https://github.com/Stewori/JyNI Sadly people just want to hack something together in Clojure often and not really solve the problems. The JyNI author is still working on this mostly alone, although it would be an enormous benefit to the JVM ecosystem.

To build a higher-level scientific pipeline autodifferentiation done in pure Clojure would be a next step after Dragan's work imo. At least for deep learning and things like variational inference in Bayesian statistics. I think it should have a pluggable interface to tensor libraries (protocols) and then even if it is naive it will provide value now and in the long run by addition of optimization transformations (like in the Clojure analyzer).  Then something like keras/cortex could be built on top of it without spoiling the composability necessary for research and exploratory programming.

On top of autodifferentiation other high-perf libraries can be built. So I think if clojure-tensorflow can factor tensorflow as just being one backend it could play an important role, even if it is only naive for now and just solves our problems.

**Comment by whilo** [comment]
Regarding communication channel, Christopher Small and me had some discussions in this direction here: https://gitter.im/metasoarous/clojure-datascience You could also create a separate channel.

**Comment by whilo** [comment]
What is your take on tensorflow/scientific computing in Clojure?

**Comment by whilo** [comment]
I have added the `sum` gradient by a constant 1. op, but it does not work in another example where I get shape mismatches. I am not sure how to debug the gradient calculation yet, at the moment it is trial and error. Is there a way to receive the whole gradient object? 
I have also investigated automatic differentiation again yesterday and played around with [sicmutils](https://github.com/littleredcomputer/sicmutils). I managed to get a hacked symbolic gradient of a small network with one sigmoid hidden layer with the `D` operator:
~~~clojure
(defn sigm [x]
  (/ 1 (+ 1 (exp (- x)))))

(with-literal-functions [i1 i2 o1 o2]
  ((->>
    ((D (fn [w11 w12 w21 w22
             v11 v12 v21 v22]
          (let [h1* (* (m/by-rows [i1 i2])
                       (m/by-rows [w11 w12]
                                  [w21 w22]))
                h11 (sigm (m/get-in h1* [0 0]))
                h12 (sigm (m/get-in h1* [0 1]))]
            (* (m/by-rows [h11 h12])
               (m/by-rows [v11 v12]
                          [v21 v22])
               [o1 o2])))) 3 5 7 11 13 17 19 23) ;; i am not sure how form the derivative with symbolic weight values yet
    )
   8)) ;; this is just to make it print

;; =>
(down
 (up
  (/
   (+
    (* 13 (i1 8) (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) (o1 8))
    (* 17 (i1 8) (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) (o2 8)))
   (+
    (expt (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) 2)
    (* 2 (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))))
    1)))
 (up
  (/
   (+
    (* 19 (i1 8) (o1 8) (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))))
    (* 23 (i1 8) (o2 8) (exp (+ (* -5 (i1 8)) (* -11 (i2 8))))))
   (+
    (expt (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))) 2)
    (* 2 (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))))
    1)))
 (up
  (/
   (+
    (* 13 (i2 8) (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) (o1 8))
    (* 17 (i2 8) (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) (o2 8)))
   (+
    (expt (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) 2)
    (* 2 (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))))
    1)))
 (up
  (/
   (+
    (* 19 (i2 8) (o1 8) (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))))
    (* 23 (i2 8) (o2 8) (exp (+ (* -5 (i1 8)) (* -11 (i2 8))))))
   (+
    (expt (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))) 2)
    (* 2 (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))))
    1)))
 (up (/ (o1 8) (+ (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) 1)))
 (up (/ (o2 8) (+ (exp (+ (* -3 (i1 8)) (* -7 (i2 8)))) 1)))
 (up (/ (o1 8) (+ (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))) 1)))
 (up (/ (o2 8) (+ (exp (+ (* -5 (i1 8)) (* -11 (i2 8)))) 1))))
~~~
The scimutils toolbox has optimization routines as well and is in general very powerful. But I am not sure how well this can be mapped to tensorflow yet. It calculates scalar derivatives. There is also https://github.com/log0ymxm/clj-auto-diff, which you probably had a look at. What is your take on modelling auto-differentiation?

**Comment by kieranbrowne** [comment]
I like your vision of a higher-level scientific pipeline for autodifferentiation. If we can set it up so its spits out operations maps, like [tools.analyser](https://github.com/clojure/tools.analyzer), we should be able to hook it up to a tensorflow backend without trouble.
```clojure
clojure.tools.analyzer.jvm> (ast/children (analyze '(do 1 2 :foo)))
[{:op   :const,
  :id   0,
  :type :number,
  :val  1,
  :form 1,
  ...}
 {:op   :const,
  :id   1,
  :type :number,
  :val  2,
  :form 2,
  ...}
 {:op   :const,
  :id   3,
  :type :keyword,
  :val  :foo,
  :form :foo,
  ...}]
```
To be honest the actual mechanics of autodifferentiation are pretty mysterious to me. My current implementation just chains together operation specific gradients from a lookup table, but there's probably a better way. Your example code above is interesting, but I'm not sure I understand the output.

You can get the gradient tf operation with:
```clojure
(let [x (tf/constant 1.)
      y (tf/mult x (tf/constant 1.))]
   (clojure-tensorflow.gradients/gradient y x)
  )
;;=> #object[org.tensorflow.Output 0x1805447d org.tensorflow.Output@1805447d]
```

But you can also get the individual paths between two operations like so.
```clojure
(let [x (tf/constant 1.)
      y (tf/mult x (tf/constant 1.))
      z (tf/mult y x)]
  (map (fn [path]
         (map (comp #(.name (.op %)) :output) path))
       (clojure-tensorflow.gradients/paths z x)))
;; => ((Mul45602 Mul45601 Const45603) (Mul45602 Const45604))
```
The above uses `clojure-tensorflow.gradients/paths` to get all reference paths between z and x in the graph. We can work out the gradient of z with respect to x by applying the chain rule for between the nodes of each path, and summing all the possible paths.

Like I said, this works, but might be naive because it requires a look up table for every low-level op, of which there are many. If you want to try out alternative methods for autodifferentiation that would be sweet.
